### PR TITLE
PCA online oneCCL sample

### PR DESCRIPTION
**Pull request by mchernov-intel**
correctness of work was checked

### CODE REVIEW

File: samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+/* Input data set parameters */
+const string dataFileNames[4] = { "./data/pca_normalized_1.csv", "./data/pca_normalized_2.csv", "./data/pca_normalized_3.csv",
+                                  "./data/pca_normalized_4.csv" };
+
+#define ccl_root 0
+
+int getLocalRank(ccl::communicator & comm, int size, int rank)
+{
+    /* Obtain local rank among nodes sharing the same host name */
+    char zero = static_cast<char>(0);
```
Comment (KulikovNikita) [inline samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp:57]:
```suggestion
    const char zero = '\0';
```

File: samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    std::vector<char> name(MPI_MAX_PROCESSOR_NAME + 1, zero);
+    int resultlen = 0;
+    MPI_Get_processor_name(name.data(), &resultlen);
+    std::string str(name.begin(), name.end());
+    std::vector<char> allNames((MPI_MAX_PROCESSOR_NAME + 1) * size, zero);
+    std::vector<size_t> aReceiveCount(size, MPI_MAX_PROCESSOR_NAME + 1);
+    ccl::allgatherv((int8_t *)name.data(), name.size(), (int8_t *)allNames.data(), aReceiveCount, comm).wait();
```
Comment (KulikovNikita) [inline samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp:64]:
```suggestion
    ccl::allgatherv(static_cast<int8_t*>(name.data()), name.size(), static_cast<int8_t*>(allNames.data()), aReceiveCount, comm).wait();
```

File: samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+const string dataFileNames[4] = { "./data/pca_normalized_1.csv", "./data/pca_normalized_2.csv", "./data/pca_normalized_3.csv",
+                                  "./data/pca_normalized_4.csv" };
+
+#define ccl_root 0
+
+int getLocalRank(ccl::communicator & comm, int size, int rank)
+{
+    /* Obtain local rank among nodes sharing the same host name */
+    char zero = static_cast<char>(0);
+    std::vector<char> name(MPI_MAX_PROCESSOR_NAME + 1, zero);
```
Comment (KulikovNikita) [inline samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp:58]: I believe ccl library should provide specific macro/const variable for this. Isn't it?

File: samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    }
+    aReceiveCount[ccl_root] = totalArchLength;
+
+    serializedData.resize(totalArchLength);
+
+    ByteBuffer nodeResults(perNodeArchLength);
+    dataArch.copyArchiveToArray(&nodeResults[0], perNodeArchLength);
+
+    /* Transfer partial results to step 2 on the root node */
+    ccl::allgatherv((int8_t *)&nodeResults[0], perNodeArchLength, (int8_t *)&serializedData[0], aPerNodeArchLength, comm).wait();
```
Comment (KulikovNikita) [inline samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp:161]:
```suggestion
    ccl::allgatherv(static_cast<int8_t*>(nodeResults), perNodeArchLength, static_cast<int8_t*>(serializedData), aPerNodeArchLength, comm).wait();
```

File: samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        MPI_Bcast((void *)main_addr.data(), main_addr.size(), MPI_BYTE, 0, MPI_COMM_WORLD);
+        kvs = ccl::create_kvs(main_addr);
+    }
+
+    auto comm = ccl::create_communicator(size, rank, kvs);
+
+    /* Create GPU device from local rank and set execution context */
+    auto local_rank = getLocalRank(comm, size, rank);
+    auto gpus       = get_gpus();
+    auto rank_gpu   = gpus[local_rank % gpus.size()];
```
Comment (KulikovNikita) [inline samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp:114]: This variables can be marked as `const`. Aren't they?

**Review by KulikovNikita — APPROVED** [review]

**Review by makart19 — APPROVED** [review]

File: samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp (hunk 1 above)
Comment (mchernov-intel) [reply samples/daal/cpp/oneccl/sources/pca_dense_online_distributed_oneccl.cpp:57]: let's move this changes to other pull request cause all oneCCL samples should be fixed.

**Pull request merged**